From 541a04ad7e7e147b51db8d6cad1fcfcc73be04e1 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Sat, 9 Aug 2025 16:28:08 +0200
Subject: [PATCH] board: solidrun: lx2160cex7: add support for half-twins board

Compile with CONFIG_DEFAULT_DEVICE_TREE=fsl-lx2160a-half-twins

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 .../dts/fsl-lx2160a-half-twins-u-boot.dtsi    |   7 +
 arch/arm/dts/fsl-lx2160a-half-twins.dts       | 584 ++++++++++++++++++
 configs/lx2160acex7_tfa_defconfig             |   1 +
 3 files changed, 592 insertions(+)
 create mode 100644 arch/arm/dts/fsl-lx2160a-half-twins-u-boot.dtsi
 create mode 100644 arch/arm/dts/fsl-lx2160a-half-twins.dts

diff --git a/arch/arm/dts/fsl-lx2160a-half-twins-u-boot.dtsi b/arch/arm/dts/fsl-lx2160a-half-twins-u-boot.dtsi
new file mode 100644
index 00000000000..25031b98a7a
--- /dev/null
+++ b/arch/arm/dts/fsl-lx2160a-half-twins-u-boot.dtsi
@@ -0,0 +1,7 @@
+// SPDX-License-Identifier: GPL-2.0+
+
+#include "fsl-lx2160a-cex7-u-boot.dtsi"
+
+&usb1 {
+	status = "disabled";
+};
diff --git a/arch/arm/dts/fsl-lx2160a-half-twins.dts b/arch/arm/dts/fsl-lx2160a-half-twins.dts
new file mode 100644
index 00000000000..497d56036e2
--- /dev/null
+++ b/arch/arm/dts/fsl-lx2160a-half-twins.dts
@@ -0,0 +1,584 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+//
+// Device Tree file for LX2160A based half twins
+//
+// Copyright 2022 SolidRun Ltd.
+
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+#include "fsl-lx2160a-cex7.dtsi"
+
+/ {
+	model = "SolidRun LX2160A Twins";
+	compatible = "solidrun,clearfog-twins",
+		"solidrun,lx2160a-cex7", "fsl,lx2160a";
+
+	aliases {
+		i2c0 = &i2c0;
+		i2c1 = &i2c2;
+		i2c2 = &i2c4;
+		i2c3 = &{/i2c@2000000/i2c-mux@77/i2c@1};
+		i2c4 = &{/i2c@2000000/i2c-mux@77/i2c@2};
+		i2c5 = &{/i2c@2000000/i2c-mux@77/i2c@3};
+		i2c6 = &sfp0_i2c;
+		i2c7 = &sfp1_i2c;
+		i2c8 = &sfp2_i2c;
+		i2c9 = &sfp3_i2c;
+		i2c10 = &twins_sfp_c1_at_i2c;
+		i2c11 = &twins_sfp_c1_ab_i2c;
+		i2c12 = &twins_sfp_c1_bt_i2c;
+		i2c13 = &twins_sfp_c1_bb_i2c;
+		i2c14 = &twins_sfp_c2_at_i2c;
+		i2c15 = &twins_sfp_c2_ab_i2c;
+		i2c16 = &twins_sfp_c2_bt_i2c;
+		i2c17 = &twins_sfp_c2_bb_i2c;
+		i2c18 = &twins_sfp_c3_at_i2c;
+		i2c19 = &twins_sfp_c3_ab_i2c;
+		i2c20 = &twins_sfp_c3_bt_i2c;
+		i2c21 = &twins_sfp_c3_bb_i2c;
+		i2c22 = &htwins_sfp_c3_at_i2c;
+		i2c23 = &htwins_sfp_c3_ab_i2c;
+		i2c24 = &htwins_sfp_c3_bt_i2c;
+		i2c25 = &htwins_sfp_c3_bb_i2c;
+		i2c26 = &{/i2c@2000000/i2c-mux@77/i2c@0};
+		mmc0 = &esdhc0;
+		mmc1 = &esdhc1;
+		serial0 = &uart0;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	c1_at_sfp: c1-at-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c1_at_i2c>;
+		mod-def0-gpio = <&expander0 1 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c1_ab_sfp: c1-ab-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c1_ab_i2c>;
+		mod-def0-gpio = <&expander0 2 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c1_bt_sfp: c1-bt-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c1_bt_i2c>;
+		mod-def0-gpio = <&expander0 3 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c1_bb_sfp: c1-bb-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c1_bb_i2c>;
+		mod-def0-gpio = <&expander0 4 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c2_at_sfp: c2-at-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c2_at_i2c>;
+		mod-def0-gpio = <&expander0 5 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c2_ab_sfp: c2-ab-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c2_ab_i2c>;
+		mod-def0-gpio = <&expander0 6 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c2_bt_sfp: c2-bt-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c2_bt_i2c>;
+		mod-def0-gpio = <&expander0 9 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c2_bb_sfp: c2-bb-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c2_bb_i2c>;
+		mod-def0-gpio = <&expander0 10 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c3_at_sfp: c3-at-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c3_at_i2c>;
+		mod-def0-gpio = <&expander0 11 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c3_ab_sfp: c3-ab-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c3_ab_i2c>;
+		mod-def0-gpio = <&expander0 12 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c3_bt_sfp: c3-bt-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c3_bt_i2c>;
+		mod-def0-gpio = <&expander0 13 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	c3_bb_sfp: c3-bb-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&twins_sfp_c3_bb_i2c>;
+		mod-def0-gpio = <&expander0 14 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	ht_c3_at_sfp: ht-c3-at-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&htwins_sfp_c3_at_i2c>;
+		mod-def0-gpio = <&expander2 11 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	ht_c3_ab_sfp: ht-c3-ab-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&htwins_sfp_c3_ab_i2c>;
+		mod-def0-gpio = <&expander2 12 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	ht_c3_bt_sfp: ht-c3-bt-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&htwins_sfp_c3_bt_i2c>;
+		mod-def0-gpio = <&expander2 13 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	ht_c3_bb_sfp: ht-c3-bb-sfp {
+		compatible = "sff,sfp";
+		i2c-bus = <&htwins_sfp_c3_bb_i2c>;
+		mod-def0-gpio = <&expander2 14 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_c1_at: led-c1-at {
+			gpios = <&expander1 1 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c1_ab: led-c1-ab {
+			gpios = <&expander1 2 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c1_bt: led-c1-bt {
+			gpios = <&expander1 4 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c1_bb: led-c1-bb {
+			gpios = <&expander1 3 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c2_at: led-c2-at {
+			gpios = <&expander1 5 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c2_ab: led-c2-ab {
+			gpios = <&expander1 6 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c2_bt: led-c2-bt {
+			gpios = <&expander1 10 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c2_bb: led-c2-bb {
+			gpios = <&expander1 9 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c3_at: led-c3-at {
+			gpios = <&expander1 11 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c3_ab: led-c3-ab {
+			gpios = <&expander1 12 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c3_bt: led-c3-bt {
+			gpios = <&expander1 14 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_c3_bb: led-c3-bb {
+			gpios = <&expander1 13 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_ht_c3_at: led-ht-c3-at {
+			gpios = <&expander3 11 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_ht_c3_ab: led-ht-c3-ab {
+			gpios = <&expander3 12 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_ht_c3_bt: led-ht-c3-bt {
+			gpios = <&expander3 14 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		led_ht_c3_bb: led-ht-c3-bb {
+			gpios = <&expander3 13 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+	};
+};
+
+&i2c2 {
+	i2c-switch@76 {
+		compatible = "nxp,pca9547";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0x76>;
+		i2c-mux-idle-disconnect;
+
+		twins_sfp_c1_at_i2c: i2c@1 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <1>;
+		};
+
+		twins_sfp_c1_ab_i2c: i2c@2 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <2>;
+		};
+
+		twins_sfp_c1_bt_i2c: i2c@3 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <3>;
+		};
+
+		twins_sfp_c1_bb_i2c: i2c@4 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <4>;
+		};
+
+		twins_sfp_c2_at_i2c: i2c@5 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <5>;
+		};
+
+		twins_sfp_c2_ab_i2c: i2c@6 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <6>;
+		};
+	};
+
+	i2c-switch@77 {
+		compatible = "nxp,pca9547";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0x77>;
+		i2c-mux-idle-disconnect;
+
+		twins_sfp_c2_bt_i2c: i2c@1 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <1>;
+		};
+
+		twins_sfp_c2_bb_i2c: i2c@2 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <2>;
+		};
+
+		twins_sfp_c3_at_i2c: i2c@3 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <3>;
+		};
+
+		twins_sfp_c3_ab_i2c: i2c@4 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <4>;
+		};
+
+		twins_sfp_c3_bt_i2c: i2c@5 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <5>;
+		};
+
+		twins_sfp_c3_bb_i2c: i2c@6 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <6>;
+		};
+	};
+
+	expander0: gpio-expander@20 {
+		compatible = "nxp,pca9555";
+		gpio-controller;
+		#gpio-cells = <2>;
+		reg = <0x20>;
+	};
+
+	expander1: gpio-expander@21 {
+		compatible = "nxp,pca9555";
+		gpio-controller;
+		#gpio-cells = <2>;
+		reg = <0x21>;
+	};
+
+	/* Half twins configuration; take over c3 from the other twin side */
+	i2c-switch@73 {
+		compatible = "nxp,pca9547";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0x73>;
+		i2c-mux-idle-disconnect;
+
+		htwins_sfp_c3_at_i2c: i2c@3 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <3>;
+		};
+
+		htwins_sfp_c3_ab_i2c: i2c@4 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <4>;
+		};
+
+		htwins_sfp_c3_bt_i2c: i2c@5 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <5>;
+		};
+
+		htwins_sfp_c3_bb_i2c: i2c@6 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <6>;
+		};
+	};
+
+	expander2: gpio-expander@24 {
+		compatible = "nxp,pca9555";
+		gpio-controller;
+		#gpio-cells = <2>;
+		reg = <0x24>;
+	};
+
+	expander3: gpio-expander@25 {
+		compatible = "nxp,pca9555";
+		gpio-controller;
+		#gpio-cells = <2>;
+		reg = <0x25>;
+	};
+};
+
+/* SD1 lanes #0.. #7 */
+&dpmac3 {
+	sfp = <&c1_at_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_c1_at>;
+};
+
+&dpmac4 {
+	sfp = <&c1_bt_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_c1_bt>;
+};
+
+&dpmac5 {
+	sfp = <&ht_c3_bt_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_ht_c3_bt>;
+};
+
+&dpmac6 {
+	sfp = <&ht_c3_at_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_ht_c3_at>;
+};
+
+&dpmac7 {
+	sfp = <&c2_at_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_c2_at>;
+};
+
+&dpmac8 {
+	sfp = <&c2_bt_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_c2_bt>;
+};
+
+&dpmac9 {
+	sfp = <&c3_at_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_c3_at>;
+};
+
+&dpmac10 {
+	sfp = <&c3_bt_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_c3_bt>;
+};
+
+/* SD2 lanes #0.. #7 */
+&dpmac11 {
+	sfp = <&ht_c3_ab_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_ht_c3_ab>;
+};
+
+&dpmac12 {
+	sfp = <&c1_ab_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_c1_ab>;
+};
+
+&dpmac13 { // ok
+	sfp = <&c3_ab_sfp>;
+	managed = "in-band-status";
+	phy-mode = "sgmii";
+	link-status-led = <&led_c3_ab>;
+};
+
+&dpmac14 { // ok
+	sfp = <&c3_bb_sfp>;
+	managed = "in-band-status";
+	phy-mode = "sgmii";
+	link-status-led = <&led_c3_bb>;
+};
+
+&dpmac15 {
+	sfp = <&ht_c3_bb_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_ht_c3_bb>;
+};
+
+&dpmac16 {
+	sfp = <&c2_bb_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_c2_bb>;
+};
+
+&dpmac17 {
+	/delete-property/ phy-handle;
+	/delete-property/ phy-connection-type;
+	sfp = <&c1_bb_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_c1_bb>;
+};
+
+&dpmac18 {
+	sfp = <&c2_ab_sfp>;
+	managed = "in-band-status";
+	link-status-led = <&led_c2_ab>;
+};
+
+&esdhc0 {
+	sd-uhs-sdr104;
+	sd-uhs-sdr50;
+	sd-uhs-sdr25;
+	sd-uhs-sdr12;
+	no-1-v-8;
+	status = "okay";
+};
+
+&pcs_mdio3 {
+	status = "okay";
+};
+
+&pcs_mdio4 {
+	status = "okay";
+};
+
+&pcs_mdio5 {
+	status = "okay";
+};
+
+&pcs_mdio6 {
+	status = "okay";
+};
+
+&pcs_mdio7 {
+	status = "okay";
+};
+
+&pcs_mdio8 {
+	status = "okay";
+};
+
+&pcs_mdio9 {
+	status = "okay";
+};
+
+&pcs_mdio10 {
+	status = "okay";
+};
+
+&pcs_mdio11 {
+	status = "okay";
+};
+
+&pcs_mdio12 {
+	status = "okay";
+};
+
+&pcs_mdio13 {
+	status = "okay";
+};
+
+&pcs_mdio14 {
+	status = "okay";
+};
+
+&pcs_mdio15 {
+	status = "okay";
+};
+
+&pcs_mdio16 {
+	status = "okay";
+};
+
+&pcs_mdio17 {
+	status = "okay";
+};
+
+&pcs_mdio18 {
+	status = "okay";
+};
+
+&uart0 {
+	status = "okay";
+};
+
+&usb0 {
+	status = "okay";
+};
diff --git a/configs/lx2160acex7_tfa_defconfig b/configs/lx2160acex7_tfa_defconfig
index 9b8a99f868b..64432d9f76b 100644
--- a/configs/lx2160acex7_tfa_defconfig
+++ b/configs/lx2160acex7_tfa_defconfig
@@ -62,6 +62,7 @@ CONFIG_DYNAMIC_DDR_CLK_FREQ=y
 CONFIG_DDR_ECC=y
 CONFIG_ECC_INIT_VIA_DDRCONTROLLER=y
 CONFIG_SYS_FSL_DDR_INTLV_256B=y
+CONFIG_DM_PCA953X=y
 CONFIG_MPC8XXX_GPIO=y
 CONFIG_DM_I2C=y
 CONFIG_I2C_SET_DEFAULT_BUS_NUM=y
-- 
2.43.0

